compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -323,151 +323,185 @@ generateS390Compare(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::Mne
- */
-static TR::Register* maxMinHelper(TR::Node* node, TR::CodeGenerator* cg, bool isMax)
+static TR::Register*
+xmaxxminHelper(TR::Node* node, TR::CodeGenerator* cg, TR::InstOpCode::Mnemonic compareRROp, TR::InstOpCode::S390BranchCondition branchCond, TR::InstOpCode::Mnemonic moveRROp)
    {
    TR::Node* lhsNode = node->getChild(0);
    TR::Node* rhsNode = node->getChild(1);
 
-   TR::Register* lhsReg = NULL;
+   TR::Register* lhsReg = cg->gprClobberEvaluate(lhsNode);
    TR::Register* rhsReg = cg->evaluate(rhsNode);
 
-   // Mask is 4 to pick rhs when lhs is less for max, 2 to pick rhs when lhs is greater for min
-   const uint8_t mask = isMax ? 0x4 : 0x2;
+   TR::LabelSymbol* cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol* cFlowRegionEnd = generateLabelSymbol(cg);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+   cFlowRegionStart->setStartInternalControlFlow();
+
+   generateRREInstruction(cg, compareRROp, node, lhsReg, rhsReg);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, branchCond, node, cFlowRegionEnd);
 
-   if (node->getOpCodeValue() == TR::imax || node->getOpCodeValue() == TR::imin)
+   generateRREInstruction(cg, moveRROp, node, lhsReg, rhsReg);
+
+   TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
+
+   deps->addPostConditionIfNotAlreadyInserted(lhsReg, TR::RealRegister::AssignAny);
+   deps->addPostConditionIfNotAlreadyInserted(rhsReg, TR::RealRegister::AssignAny);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionEnd, deps);
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   node->setRegister(lhsReg);
+
+   cg->decReferenceCount(lhsNode);
+   cg->decReferenceCount(rhsNode);
+
+   return lhsReg;
+   }
+
+static TR::Register*
+imaximinHelper(TR::Node* node, TR::CodeGenerator* cg, TR::InstOpCode::Mnemonic compareRROp, TR::InstOpCode::S390BranchCondition branchCond, TR::InstOpCode::Mnemonic moveRROp)
+   {
+   if (cg->comp()->target().cpu.getSupportsArch(TR::CPU::z15))
       {
-      if (cg->comp()->target().cpu.getSupportsArch(TR::CPU::z15))
-         {
-         lhsReg = cg->allocateRegister();
+      TR::Node* lhsNode = node->getChild(0);
+      TR::Node* rhsNode = node->getChild(1);
 
-         // Load into a tmp instead of clobberEvaluating into lhsReg to avoid an extra register shuffle
-         TR::Register* tmpRegister = cg->evaluate(lhsNode);
+      TR::Register* resultReg = cg->allocateRegister();
 
-         generateRRInstruction(cg, TR::InstOpCode::CR, node, tmpRegister, rhsReg);
-         generateRRFInstruction(cg, TR::InstOpCode::SELR, node, lhsReg, rhsReg, tmpRegister, mask);
-         }
-      else if (cg->comp()->target().cpu.getSupportsArch(TR::CPU::z196))
-         {
-         lhsReg = cg->gprClobberEvaluate(lhsNode);
+      // Load into a tmp instead of clobberEvaluating into resultReg to avoid an extra register shuffle
```

Comment:
Can you explain the shuffling part of this comment please?  If one of the LHS or RHS nodes dies here, what is the consequence of using the "dead" register as the resultReg?  Or is that not allowed in the `SELR` instruction?

Reply:
These comments are now stale. Thanks for catching them. I've removed them in a [force-push](https://github.com/eclipse/omr/compare/9efde58c34032acb4190ab6b049ea5ede7dddbaf..2ae63d4a98195d0c905705614b11eba616399fbf).

To explain why they were originally there, looking at the original code before the change [1], we do not call `clobberEvaluate` because there may be other uses of the register associated with the node, and thus the `clobberEvaluate` would have to generate a register-register shuffle following the evaluation. This extra register shuffle is not needed because the `SELR` (select register) instruction will overwrite the target anyway.

For example, using `clobberEvaluate` may have resulted in the following sequence:

```
LR GPR_TMP, GPR_LHS
CR GPR_TMP, GPR_RHS
SELR GPR_TMP, GPR_LHS, GPR_RHS
```

Which is inefficient because the `LR` instruction is redundant. Allocating the register by ourselves avoids the extra `LR` shuffle at no cost, because the `SELR` instruction will define the target register, and if it so happens that the use of `GPR_LHS` is the last use of the register in the instruction sequence, then `GPR_TMP` will become unlatched as we are register allocating and local RA can pick the same real register for both `GPR_TMP` and `GPR_LHS`.

[1] https://github.com/eclipse/omr/blob/00c57af16de56a1458194f254fb76786220b7bfe/compiler/z/codegen/ControlFlowEvaluator.cpp#L358-L364

Reply:
@0xdaryl does the above answer your question? Any other concerns?

Reply:
@0xdaryl any other concerns?